### PR TITLE
Add a MailRu provider

### PR DESCRIPTION
**Pull request by dyzy**
New provider for MailRu (www.mail.ru)
Details from provider: https://o2.mail.ru/docs#web

### CODE REVIEW

File: src/AspNet.Security.OAuth.MailRu/MailRuAuthenticationDefaults.cs [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+ * for more information concerning the license and the contributors participating to this project.
+ */
+
+using Microsoft.AspNetCore.Authentication;
+using Microsoft.AspNetCore.Authentication.OAuth;
+
+namespace AspNet.Security.OAuth.MailRu
+{
+    /// <summary>
+    /// Default values used by the GitHub authentication middleware.
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.MailRu/MailRuAuthenticationDefaults.cs]: _GitHub_ -> _MailRu_

File: src/AspNet.Security.OAuth.MailRu/MailRuAuthenticationHandler.cs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+            [NotNull] UrlEncoder encoder,
+            [NotNull] ISystemClock clock)
+            : base(options, logger, encoder, clock)
+        {
+        }
+
+        protected override async Task<AuthenticationTicket> CreateTicketAsync([NotNull] ClaimsIdentity identity,
+            [NotNull] AuthenticationProperties properties, [NotNull] OAuthTokenResponse tokens)
+        {
+            var request = new HttpRequestMessage(HttpMethod.Get, Options.UserInformationEndpoint + "?access_token=" + tokens.AccessToken);
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.MailRu/MailRuAuthenticationHandler.cs]: Use `QueryHelpers.AddQueryString(string, string, string)` to encode query string parameters.

Example:

https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/blob/af1e3373cf00e82b186ea2a4bba0696d476b0dff/src/AspNet.Security.OAuth.Vkontakte/VkontakteAuthenticationHandler.cs#L44

File: src/AspNet.Security.OAuth.MailRu/MailRuAuthenticationHandler.cs [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+            var payload = JObject.Parse(await response.Content.ReadAsStringAsync());
+
+            var principal = new ClaimsPrincipal(identity);
+            var context = new OAuthCreatingTicketContext(principal, properties, Context, Scheme, Options, Backchannel, tokens, payload);
+            context.RunClaimActions(payload);
+
+            await Options.Events.CreatingTicket(context);
+            return new AuthenticationTicket(context.Principal, context.Properties, Scheme.Name);
+        }
+
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.MailRu/MailRuAuthenticationHandler.cs]: Nit: delete this blank line.

File: src/AspNet.Security.OAuth.MailRu/MailRuAuthenticationOptions.cs [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+        {
+            ClaimsIssuer = MailRuAuthenticationDefaults.Issuer;
+
+            CallbackPath = new PathString(MailRuAuthenticationDefaults.CallbackPath);
+
+            AuthorizationEndpoint = MailRuAuthenticationDefaults.AuthorizationEndpoint;
+            TokenEndpoint = MailRuAuthenticationDefaults.TokenEndpoint;
+            UserInformationEndpoint = MailRuAuthenticationDefaults.UserInformationEndpoint;
+
+            //Scope.Add("userinfo");
```
Comment (martincostello) [inline src/AspNet.Security.OAuth.MailRu/MailRuAuthenticationOptions.cs]: Delete commented-out code.

File: test/AspNet.Security.OAuth.Providers.Tests/MailRu/MailRuTests.cs [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+            });
+        }
+
+        [Theory]
+        [InlineData(ClaimTypes.NameIdentifier, "123abc")]
+        [InlineData(ClaimTypes.Name, "Vasya")]
+        [InlineData(ClaimTypes.Email, "vasya@mail.ru")]
+        [InlineData(ClaimTypes.GivenName, "Vasiliy")]
+        [InlineData(ClaimTypes.Surname, "Ivanov")]
+        [InlineData("urn:mailru:profileimage", "https://filin.mail.ru/pic?d=idofpic&width=180&height=180")]
```
Comment (martincostello) [inline test/AspNet.Security.OAuth.Providers.Tests/MailRu/MailRuTests.cs:40]: Can you add a test case for `ClaimTypes.Gender`?

**Review by martincostello — COMMENTED** [review]
Can you also add the provider to the `README` please? There's a template commented-out at the end of the file.

https://github.com/aspnet-contrib/AspNet.Security.OAuth.Providers/blob/dev/README.md#providers

**Comment by dyzy** [comment]
Thanks for comments! Just did required changes. Please review.

**Review by martincostello — APPROVED** [review]

**Comment by martincostello** [comment]
Thanks for your contribution @dyzy! 🎈

**Comment by martincostello** [comment]
This provider has now been deployed to [NuGet.org](https://www.nuget.org/packages/AspNet.Security.OAuth.MailRu/2.1.0) - thanks again for your contribution @dyzy! 🎉

**Pull request merged**